### PR TITLE
LPS-120199 Update liferay-npm-scripts to v32.6.1

### PR DESCRIPTION
**Pull request by wincent**
- [Release notes for v32.6.0](https://github.com/liferay/liferay-npm-tools/releases/tag/liferay-npm-scripts%2Fv32.6.0).
- [Release notes for v32.6.1](https://github.com/liferay/liferay-npm-tools/releases/tag/liferay-npm-scripts%2Fv32.6.1).

Two main things of interest:

1.  Updates JS toolkit dependencies, in order to get [this PR](https://github.com/liferay/liferay-js-toolkit/pull/632) which adds "preload" to link tags to improve Lighthouse performance audit; that one will close [LPS-115876](https://issues.liferay.com/browse/LPS-115876).

2.  Only merge babel options into babel webpack loader, not other loaders; requested by Commerce folks ([related PR](https://github.com/liferay/liferay-npm-tools/pull/481)); cc @FabioDiegoMastrorilli

The second commit is just some defensive future-proofing of the configuration, which I believe shouldn't actually impact current behavior at all.

### CODE REVIEW

**Comment by wincent** [comment]
Added you as a reviewer, @izaera, to get your opinion on that second commit. I _think_ it's safe, but it is always good to have a second opinion. Some of those modules don't even have source files, and none of them have loader rules at all.

However, this is still a draft because in testing this, I saw some alarming stuff in the console:

```
Uncaught ReferenceError: define is not defined
    at HTMLLinkElement.defineModule (VM247 Modal.scss.js:22)
```

This is what the compiled ".scss.js" file looks like this in the network tab:

```js
(function () {
  var link = document.createElement("link");
  var preload;

  try {
    preload = link.relList.supports("preload");
  } catch (error) {
    preload = false;
  }

  if (preload) {
    link.setAttribute("as", "style");
    link.setAttribute("rel", "preload");
  } else {
    link.setAttribute("rel", "stylesheet");
  }

  link.setAttribute("type", "text/css");
  link.setAttribute("href", Liferay.ThemeDisplay.getPathContext() + "/o/frontend-js-web/liferay/modal/Modal.css");

  function defineModule() {
    define("frontend-js-web@4.0.27/liferay/modal/Modal.scss", ['module', 'exports', 'require'], function (module, exports, require) {
      var define = undefined;
      var global = window;
      {
        module.exports = link;
      }
    });
  }

  link.onload = defineModule;

  link.onerror = function () {
    console.warn('Unable to load /o/frontend-js-web/liferay/modal/Modal.css. However, its .js module will still be defined to avoid breaking execution flow (expect some visual degradation).');
    defineModule();
  };

  document.querySelector("head").appendChild(link);
})();
//# sourceMappingURL=Modal.scss.js.map
```

The wrapped module calls naked `define`, and it should probably be `Liferay.Loader.define`. This wasn't happening in my testing of [the PR on Friday](https://github.com/liferay/liferay-js-toolkit/pull/632) and you said you saw it working too, so I am going to investigate to see what's gone wrong. Could be something in my local setup, or could be a bug. Will report back when I know more.

**Comment by wincent** [comment]
> Will report back when I know more.

Definitely happening (not on master, happens on this PR). I am suspicious of [this commit](https://github.com/liferay/liferay-js-toolkit/commit/0c14c6970503d66d73cb5b311d15ce5748d9eca5#diff-ddf68954e6308dfdd0bc40398b4039de) to the babel-plugin-namespace-amd-define package. If I can verify that, will send a fix.

**Comment by wincent** [comment]
Proposed fix for the regression [is here](https://github.com/liferay/liferay-js-toolkit/pull/636), but I have no idea whether it is the right one, so will need @izaera to check it out (I'm worried that any "fix" I apply to cure the problem here will just wind up regressing some other use case that I don't know about).

**Comment by wincent** [comment]
Force-pushed to rebase on current upstream `HEAD` of `master`, and bumped the scripts to v32.6.1 in order to get [this fix](https://github.com/liferay/liferay-js-toolkit/pull/636).

Testing in DXP now, and going to take it out of draft in the meantime.

**Comment by wincent** [comment]
It works:

Screenshot 2020-09-02 -181300-iNTdr0uA@2x

So going to run the tests. I'll be gone by the time they finish, but if @john-co or @bryceosterhaus or someone on that side of the pond sees them green, feel free to `ci:forward`.

There is some follow-up to do as noted in [this comment](https://github.com/liferay/liferay-js-toolkit/pull/632#issuecomment-685004014):

> I haven't checked exactly why is this loading but I presume it's because we pack `Modal.js` in our global or something like that for the static `Liferay.Util.openModal`... so maybe we could also try to defer that so `Modal.js` is not loaded upfront... that would be an even better improvement

I'll create a separate Jira ticket for that and link to it from https://github.com/liferay/liferay-js-toolkit/pull/632.

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
ci:test:bundle

**Comment by john-co** [comment]
ci:forward

**Comment by john-co** [comment]
ci:stop

**Comment by john-co** [comment]
oops pulled the trigger too fast on ci:forward :sweat_smile:

**Comment by john-co** [comment]
ci:stop

**Comment by wincent** [comment]
Going to re-run tests before I start trying to analyze the failures. Will check results when I get back from errands.

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
The new failures looks unrelated, but the other three are consistent with the first run, so I am going to rebase this, re-try, and see if I can repro locally etc.

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
Failures seem quite stable. @PablitoBonito or @john-co, can you help me interpret these?

**Comment by wincent** [comment]
That [first one](https://testray.liferay.com/reports/production/logs/test-1-23/1599126161054/test-portal-acceptance-pullrequest(master)/6547/functional-tomcat90-mysql57-jdk8/3/0/LocalFile.Depot_Smoke/index.html.gz), for instance; is this just a stale selector?

Screenshot 2020-09-03 -141216-wIkIqPhB@2x

> `Element is not visible at "//div[contains(@class,'applications-menu-nav-columns')]//li[contains(.,'System')]/following-sibling::li//span[normalize-space(text())='Virtual Instances']"`

File: modules/apps/frontend-js/frontend-js-metal-web/.npmbundlerrc [hunk 1]
```diff
@@ -21,5 +21,6 @@
 		]
 	},
 	"output": "build/node/packageRunBuild/resources",
-	"preset": "liferay-npm-bundler-preset-standard"
+	"preset": "liferay-npm-bundler-preset-standard",
+	"source": ["src/main/resources/META-INF/resources"]
```
Comment (izaera) [inline modules/apps/frontend-js/frontend-js-metal-web/.npmbundlerrc:25]: I believe this is `sources` instead of `source` :point_right: https://github.com/liferay/liferay-js-toolkit/wiki/.npmbundlerrc-file-reference#sources

**Comment by izaera** [comment]
Regarding commit https://github.com/liferay-frontend/liferay-portal/pull/345/commits/36e92c308636d680aaf0fdd3ac6c27f952191b58 , it LGTM...

**Comment by wincent** [comment]
Thanks @izaera. Force-pushed with config changes amended into the second commit.

**Comment by wincent** [comment]
ci:test:relevant

**Comment by john-co** [comment]
[PR Bug]

- looking at a few of the unique failures, the are affected by the application menu being less than full width. So causing the virtual instance link is not visible on the page. There is no scroll bar in order to see the app links that are cut off
screenshot_20

**Comment by john-co** [comment]
> That first one, for instance; is this just a stale selector?

The locator is correct, selenium just doesn't see it visible on the page because the virtual instance link is cut off.

**Comment by wincent** [comment]
I have an hour or two now so going to take this up again.

**Comment by FabioDiegoMastrorilli** [comment]
@wincent can we have some time estimations for this? We are about to start rewriting many MetalJs components in React and this PR will make our static env work again :3

**Comment by jbalsas** [comment]
> We are about to start rewriting many MetalJs components in React

Please, tell me this is not for the `7.3` release 😂

**Comment by FabioDiegoMastrorilli** [comment]
It's for 6.2 :O

**Comment by jbalsas** [comment]
> It's for 6.2 :O

kidding

**Comment by FabioDiegoMastrorilli** [comment]
No worries Chema, it's for 7.4 :D but we have to rewrite half of the the components we have on the front store and it will take aaaa lot

**Comment by wincent** [comment]
> @wincent can we have some time estimations for this? We are about to start rewriting many MetalJs components in React and this PR will make our static env work again :3

I'm looking at it today (comparing JARs to see why the new toolkit produces a buggy result), but then again, this is not the first day I've looked at it, so we'll see...

**Comment by bryceosterhaus** [comment]
@FabioDiegoMastrorilli are you re-writing metal components that are in soy or jsx? If they are jsx, I can try and help you out with using https://github.com/bryceosterhaus/metal-to-react

**Comment by wincent** [comment]
TL;DR:

- Preloading isn't going to work on Chrome or Safari, but it works on Firefox.
- Going to revert the preloading change in the toolkit, cut a new release, get in some other pending change(s) to the scripts, release, and send a PR that replaces this one.

---

Started by comparing some JARs (master vs PR) using, for example `git diff -b /Users/greghurrell/Desktop/master/com.liferay.frontend.js.web-4.0.30/META-INF/resources /Users/greghurrell/Desktop/branch/com.liferay.frontend.js.web-4.0.30/META-INF/resources`:

-   Whitespace changes, which is why we're using `git diff -b`; example:

    ```diff
    --- a/Users/greghurrell/Desktop/master/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/breakpoints.js
    +++ b/Users/greghurrell/Desktop/branch/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/breakpoints.js
    @@ -1,28 +1,31 @@
     Liferay.Loader.define("frontend-js-web@4.0.30/liferay/breakpoints", ['module', 'exports', 'require'], function (module, exports, require) {
       var define = undefined;
    -  Object.defineProperty(exports, "__esModule", {
    -    value: true
    -  });
    -  exports["default"] = void 0;
    +  var global = window;
    +  {
    +    Object.defineProperty(exports, "__esModule", {
    +      value: true
    +    });
    +    exports["default"] = void 0;

    -  /**
    -   * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
    -   *
    -   * This library is free software; you can redistribute it and/or modify it under
    -   * the terms of the GNU Lesser General Public License as published by the Free
    -   * Software Foundation; either version 2.1 of the License, or (at your option)
    -   * any later version.
    -   *
    -   * This library is distributed in the hope that it will be useful, but WITHOUT
    -   * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    -   * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
    -   * details.
    -   */
    -  var _default = {
    -    PHONE: 768,
    -    TABLET: 980
    -  };
    -  exports["default"] = _default;
    -  //# sourceMappingURL=breakpoints.js.map
    +    /**
    +     * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
    +     *
    +     * This library is free software; you can redistribute it and/or modify it under
    +     * the terms of the GNU Lesser General Public License as published by the Free
    +     * Software Foundation; either version 2.1 of the License, or (at your option)
    +     * any later version.
    +     *
    +     * This library is distributed in the hope that it will be useful, but WITHOUT
    +     * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    +     * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
    +     * details.
    +     */
    +    var _default = {
    +      PHONE: 768,
    +      TABLET: 980
    +    };
    +    exports["default"] = _default;
    +    //# sourceMappingURL=breakpoints.js.map
    +  }
     });
     //# sourceMappingURL=breakpoints.js.map
    \ No newline at end of file
    ```

    ie. indentation has increased by 2 for most lines in the file due to wrapping in a block (described below).

-   AMD modules now wrapped in a block that is prepended with `var global = window` (don't think that would cause any issues); example:

    ```diff
    diff --git a/Users/greghurrell/Desktop/master/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/modal/commands/OpenSimpleInputModal.es.js b/Users/greghurrell/Desktop/branch/com.liferay.frontend.js>
    index f3e3cf18c73d..97377998dc22 100644
    --- a/Users/greghurrell/Desktop/master/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/modal/commands/OpenSimpleInputModal.es.js
    +++ b/Users/greghurrell/Desktop/branch/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/modal/commands/OpenSimpleInputModal.es.js
    @@ -1,5 +1,7 @@
     Liferay.Loader.define("frontend-js-web@4.0.30/liferay/modal/commands/OpenSimpleInputModal.es", ['module', 'exports', 'require', 'frontend-js-react-web', 'frontend-js-react-web$react', 'frontend-js-react-web$>
       var define = undefined;
    +  var global = window;
    +  {
         Object.defineProperty(exports, "__esModule", {
           value: true
         });
    @@ -106,5 +108,6 @@ Liferay.Loader.define("frontend-js-web@4.0.30/liferay/modal/commands/OpenSimpleI
         var _default = openSimpleInputModalImplementation;
         exports["default"] = _default;
         //# sourceMappingURL=OpenSimpleInputModal.es.js.map
    +  }
     });
     //# sourceMappingURL=OpenSimpleInputModal.es.js.map
    \ No newline at end of file

    ```

-   (Expected) Loader changes:

    ```diff
    diff --git a/Users/greghurrell/Desktop/master/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/modal/Modal.scss.js b/Users/greghurrell/Desktop/branch/com.liferay.frontend.js.web-4.0.30/META-INF/r>
    index 48d2c555f9db..8c188881e116 100644
    --- a/Users/greghurrell/Desktop/master/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/modal/Modal.scss.js
    +++ b/Users/greghurrell/Desktop/branch/com.liferay.frontend.js.web-4.0.30/META-INF/resources/liferay/modal/Modal.scss.js
    @@ -1,14 +1,30 @@
     (function () {
       var link = document.createElement("link");
    +  var preload;
    +
    +  try {
    +    preload = link.relList.supports("preload");
    +  } catch (error) {
    +    preload = false;
    +  }
    +
    +  if (preload) {
    +    link.setAttribute("as", "style");
    +    link.setAttribute("rel", "preload");
    +  } else {
         link.setAttribute("rel", "stylesheet");
    +  }
    +
       link.setAttribute("type", "text/css");
       link.setAttribute("href", Liferay.ThemeDisplay.getPathContext() + "/o/frontend-js-web/liferay/modal/Modal.css");

       function defineModule() {
         Liferay.Loader.define("frontend-js-web@4.0.30/liferay/modal/Modal.scss", ['module', 'exports', 'require'], function (module, exports, require) {
           var define = undefined;
    -
    +      var global = window;
    +      {
             module.exports = link;
    +      }
         });
       }
    ```

So, I had tested these loader changes manually in https://github.com/liferay/liferay-js-toolkit/pull/632 (with some difficulty). It was possible that they are responsible for regressing the application menu (which is the main visual bug I am using here to distinguish between master and the PR), and maybe I just didn't notice in my original testing — I saw the elements getting inserted and the CSS loading over the network, clicked around DXP and everything seemed fine. As the comments on that branch show, testing the toolkit in liferay-portal is tricky, so I decided to see if I could test this with some nasty hackery in `frontend-js-web/src/main/resources/META-INF/resources/liferay/global.es.js`:

```js
const proto = document.createElement('link').relList.__proto__;
const supports = proto.supports;
const value = localStorage.getItem('preloadz') ?? 'true';
proto.supports = (wat) => wat === 'preload' ? JSON.parse(value) : supports(wat);
console.log(`monkey patched: preload = ${value}`);
```

With that I can turn preloading on or off from the console without redeploying multiple packages in an error-prone way.

I ruled out things like load order. The same files are being requested, in the same order:

_With preloading (broken)_

modal-css-page-order-buggy

_Without preloading (working)_

modal-css-page-order-good

With no suspicious overlaps (as per the waterfall view):

_With preloading (broken)_

waterfall-good

_Without preloading (working)_

waterfall-buggy

Comparing styles on, for example, the `.applications-menu-wrapper` element (ie. using "Copy styles" in the chrome inspector, and then comparing good and bad versions with `git diff`):

```diff
diff --git a/good b/bad
index db97afa..31881df 100644
--- a/good
+++ b/bad
@@ -150,8 +150,4 @@ color: var(--body-color);
 font-family: var(--font-family-base);
 font-size: var(--font-size-base);
 pointer-events: auto;
-box-sizing: border-box;
-display: flex;
-flex-direction: column;
-height: -webkit-fill-available;
-max-height: 100vh;
\ No newline at end of file
+box-sizing: border-box;
\ No newline at end of file
```

So, we have missing styles from `.applications-menu-wrapper` (defined in `ApplicationsMenu.css`) but it's not in the inspector at all, even though you can see it got fetched in the network pane. 🤔

Rereading the [preloading docs on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content) suggest that maybe I didn't understand quite how `preload` is supposed to work... Their first example shows this:

```html
<link rel="preload" href="style.css" as="style">
<link rel="preload" href="main.js" as="script">

<link rel="stylesheet" href="style.css">
```

In other words, `preload` of "styles.css" (to make the browser start fetching it) _and then_ a normal `link` tag to actually apply it to the page. This is a contrived example, but it makes it look like we're doing it wrong, because we don't have that second `link` tag (and if we did, it would defeat our original purpose of trying to improve our Lighthouse metrics).

So, I tried inserting the tags by hand (in the browser console):

```js
link = document.createElement('link')
link.setAttribute('rel', 'stylesheet')
link.setAttribute('type', 'text/css')
link.setAttribute('src', '/o/frontend-js-web/liferay/modal/Modal.css')
document.head.appendChild(link)

link = document.createElement('link')
link.setAttribute('rel', 'stylesheet')
link.setAttribute('type', 'text/css')
link.setAttribute('src', '/o/product-navigation-applications-menu-web/css/ApplicationsMenu.css')
document.head.appendChild(link)
```

No effect! The styles don't get refetched (they were already fetched) _and_ they don't get applied to the DOM. At this point, I am out of theories; just a mild suspicion that this message in the console provides a clue:

> The resource http://0.0.0.0:8080/o/product-navigation-applications-menu-web/css/ApplicationsMenu.css was preloaded using link preload but not used within a few seconds from the window's load event. Please make sure it has an appropriate `as` value and it is preloaded intentionally.

My guess is that if you don't make use of a preloaded resource within a certain time window, that message gets shown and then, even if you _do_ use it (by inserting a `link` tag like I did), it's already too late and the browser ignores it. This might be a browser bug, depending on your point of view, or perhaps just undefined behavior. I didn't bother testing the theory about the time window because it would be much harder to test (might have to apply the change in the loader itself instead of a silly monkey patch), and the fact that this defeats our Lighthouse "optimization" makes it untenable anyway.

Final check: is this happening the same in different engines?

- Safari: yep; fails in the same way. So this behavior is consistent across WebKit (even significant forks like Chrome's):

safari-failing

- Firefox: nop; it works, even without the extra `link`:

firefox-working

Conclusion? Going to close this and redo as described in the TL;DR at the top of this comment.

**Comment by jbalsas** [comment]
amazingjourney

**Comment by wincent** [comment]
Replacement PR: https://github.com/liferay-frontend/liferay-portal/pull/393

**Comment by izaera** [comment]
Couldn't this be an interaction between JS and CSS? 

I mean: some JS may be modifying the applied styles and because of the preload maybe, when the JS arrives to that modification, the CSS has not been loaded yet and the JS does a different thing. Or something alike... :thinking: 

In any case, the Lighthouse warning was a false positive so it's not so horrible. Just that we need to live with that exception :shrug:

**Comment by jbalsas** [comment]
Maybe we can [file an issue in the Lighthouse repo](https://github.com/GoogleChrome/lighthouse/issues)?

**Comment by wincent** [comment]
> I mean: some JS may be modifying the applied styles and because of the preload maybe, when the JS arrives to that modification, the CSS has not been loaded yet and the JS does a different thing. 

Anything is possible, but I couldn't figure it out. The thing that I just couldn't get past is that:

1. You can see the CSS in the network pane; it's there, textually, downloaded with the right MIME type and everything.
2. You don't see it in the DOM inspector, meaning that although the element has class `foo` and you've downloaded CSS that targets `.foo`, it's not getting applied.
3. A textual search for `foo` shows no obvious manipulation of the element or its styles from JS, which makes the hypothesis you advance seem less likely. (Not impossible, obviously, just less likely.)

I could certainly dig deeper, but I am not sure if the juice would be worth the squeeze.

> Maybe we can [file an issue in the Lighthouse repo](https://github.com/GoogleChrome/lighthouse/issues)?

What's the relationship to Lighthouse exactly?

**Comment by jbalsas** [comment]
> > Maybe we can [file an issue in the Lighthouse repo](https://github.com/GoogleChrome/lighthouse/issues)?
> 
> What's the relationship to Lighthouse exactly?

I was referring to @izaera's comment:

> In any case, the Lighthouse warning was a false positive so it's not so horrible. Just that we need to live with that exception 🤷

Maybe we can file an issue so they explore wether this is really a false positive or some other wrong understanding on our part. If it's the first, maybe they'll fix it and if it's the latter, we take those learnings home ;)

**Comment by izaera** [comment]
> What's the relationship to Lighthouse exactly?

When I did the first analysis I noticed that, even without using `preload`, the CSS was loaded way before it was used, but Lighthouse still insisted that maybe we could benefit from `preload`. My guess was that it was more like a recommendation/hint than something Lighthouse found very problematic. But, if we are going to file a bug, we may need to analyze it more in-depth and motivate it, I suppose.

**Comment by jbalsas** [comment]
My understanding was that `lighthouse` was flagging a blocking link in `head` without `preload` that wasn't being used.

What we thought was a bug was the fact that the `link` was dynamically added by some async JS script, thus it couldn't have blocked initial rendering. In essence, we're implementing a `preload` type of behaviour that Lighthouse was flagging incorrectly.

Feel free to just drop this if it doesn't make sense :D

**Comment by wincent** [comment]
> But, if we are going to file a bug, we may need to analyze it more in-depth and motivate it, I suppose.

And provide them with a minimal repro recipe... "Download this 1.2 GB bundle archive and execute it..."

But seriously though, if a smaller repro could actually be made, filing a bug wouldn't be a bad idea.

**Comment by izaera** [comment]
For the record: https://github.com/GoogleChrome/lighthouse/issues?q=is%3Aissue+is%3Aopen+preload

**Comment by izaera** [comment]
> My understanding was that lighthouse was flagging a blocking link in head without preload that wasn't being used.
> ...

It is probably that. My :brain: forgets things too quickly...
